File: sabot/operators/registry.py
#!/usr/bin/env python3
"""
Operator Registry

Central registry for all physical operators in Sabot.
Provides single source of truth for operator implementations.
"""

import logging
from typing import Dict, Type, Any, List, Optional, Callable

logger = logging.getLogger(__name__)


class OperatorRegistry:
    """
    Central registry for physical operators.
    
    All engines (Stream, SQL, Graph) register their operators here,
    ensuring single implementation and consistent behavior.
    
    Example:
        registry = OperatorRegistry()
        registry.register('group_by', CythonGroupByOperator)
        
        # Create operator instance
        op = registry.create('group_by', source, ['key'])
    """
    
    def __init__(self):
        self._operators: Dict[str, Type] = {}
        self._factories: Dict[str, Callable] = {}
        self._metadata: Dict[str, Dict[str, Any]] = {}
    
    def register(
        self,
        name: str,
        operator_class: Type,
        factory: Optional[Callable] = None,
        metadata: Optional[Dict[str, Any]] = None
    ):
        """
        Register an operator.
        
        Args:
            name: Operator name (e.g., 'group_by', 'hash_join')
            operator_class: Operator class
            factory: Optional factory function for custom creation
            metadata: Optional metadata (description, performance characteristics)
        """
        if name in self._operators:
            logger.warning(f"Overwriting existing operator: {name}")
        
        self._operators[name] = operator_class
        
        if factory:
            self._factories[name] = factory
        
        if metadata:
            self._metadata[name] = metadata
        
        logger.debug(f"Registered operator: {name} ({operator_class.__name__})")
    
    def get(self, name: str) -> Optional[Type]:
        """
        Get operator class by name.
        
        Args:
            name: Operator name
            
        Returns:
            Operator class or None if not found
        """
        return self._operators.get(name)
    
    def create(self, name: str, *args, **kwargs) -> Any:
        """
        Create operator instance.
        
        Args:
            name: Operator name
            *args: Positional arguments for operator constructor
            **kwargs: Keyword arguments for operator constructor
            
        Returns:
            Operator instance
            
        Raises:
            ValueError: If operator not found
        """
        if name not in self._operators:
            raise ValueError(
                f"Unknown operator: {name}. "
                f"Available: {', '.join(self.list_operators())}"
            )
        
        # Use factory if available
        if name in self._factories:
            return self._factories[name](*args, **kwargs)
        
        # Otherwise use constructor
        return self._operators[name](*args, **kwargs)
    
    def list_operators(self) -> List[str]:
        """
        List all registered operators.
        
        Returns:
            List of operator names
        """
        return list(self._operators.keys())
    
    def get_metadata(self, name: str) -> Optional[Dict[str, Any]]:
        """
        Get metadata for an operator.
        
        Args:
            name: Operator name
            
        Returns:
            Metadata dict or None
        """
        return self._metadata.get(name)
    
    def has_operator(self, name: str) -> bool:
        """
        Check if operator is registered.
        
        Args:
            name: Operator name
            
        Returns:
            True if operator exists
        """
        return name in self._operators
    
    def unregister(self, name: str):
        """
        Remove an operator from registry.
        
        Args:
            name: Operator name
        """
        if name in self._operators:
            del self._operators[name]
            logger.debug(f"Unregistered operator: {name}")
        
        if name in self._factories:
            del self._factories[name]
        
        if name in self._metadata:
            del self._metadata[name]
    
    def clear(self):
        """Clear all registered operators."""
        self._operators.clear()
        self._factories.clear()
        self._metadata.clear()
        logger.debug("Cleared operator registry")


def create_default_registry() -> OperatorRegistry:
    """
    Create operator registry with all default operators.
    
    Registers operators from:
    - sabot._cython.operators (Cython implementations)
    - sabot_sql operators (SQL-specific)
    - sabot_cypher operators (Graph-specific)
    
    Returns:
        OperatorRegistry: Registry with default operators
    """
    registry = OperatorRegistry()
    
    # Register transform operators
    try:
        from sabot._cython.operators.transform import (
            CythonFilterOperator,
            CythonMapOperator,
            CythonSelectOperator,
            CythonFlatMapOperator,
            CythonUnionOperator
        )
        
        registry.register('filter', CythonFilterOperator, metadata={
            'type': 'transform',
            'description': 'Filter rows based on predicate',
            'performance': '10-500M rows/sec (SIMD)'
        })
        
        registry.register('map', CythonMapOperator, metadata={
            'type': 'transform',
            'description': 'Transform rows with function',
            'performance': '10-100M rows/sec'
        })
        
        registry.register('select', CythonSelectOperator, metadata={
            'type': 'transform',
            'description': 'Project columns (zero-copy)',
            'performance': '50-1000M rows/sec'
        })
        
        registry.register('flat_map', CythonFlatMapOperator, metadata={
            'type': 'transform',
            'description': '1-to-N row expansion',
            'performance': '5-50M rows/sec'
        })
        
        registry.register('union', CythonUnionOperator, metadata={
            'type': 'transform',
            'description': 'Merge multiple streams',
            'performance': '5-50M rows/sec'
        })
        
        logger.info("Registered transform operators")
        
    except ImportError as e:
        logger.warning(f"Failed to import transform operators: {e}")
    
    # Register aggregation operators
    try:
        from sabot._cython.operators.aggregations import (
            CythonGroupByOperator,
            CythonAggregateOperator,
            CythonReduceOperator,
            CythonDistinctOperator
        )
        
        registry.register('group_by', CythonGroupByOperator, metadata={
            'type': 'aggregation',
            'description': 'Group by keys and aggregate',
            'performance': '5-100M rows/sec'
        })
        
        registry.register('aggregate', CythonAggregateOperator, metadata={
            'type': 'aggregation',
            'description': 'Global aggregation (sum, mean, etc.)',
            'performance': '2-50M rows/sec'
        })
        
        registry.register('reduce', CythonReduceOperator, metadata={
            'type': 'aggregation',
            'description': 'Custom reduction function',
            'performance': '1-10M rows/sec'
        })
        
        registry.register('distinct', CythonDistinctOperator, metadata={
            'type': 'aggregation',
            'description': 'Deduplication',
            'performance': '1-10M rows/sec'
        })
        
        logger.info("Registered aggregation operators")
        
    except ImportError as e:
        logger.warning(f"Failed to import aggregation operators: {e}")
    
    # Register join operators
    try:
        from sabot._cython.operators.joins import (
            CythonHashJoinOperator,
            CythonIntervalJoinOperator,
            CythonAsofJoinOperator
        )
        
        registry.register('hash_join', CythonHashJoinOperator, metadata={
            'type': 'join',
            'description': 'Hash join on keys',
            'performance': '0.15-0.17M rows/sec (build+probe)'
        })
        
        registry.register('interval_join', CythonIntervalJoinOperator, metadata={
            'type': 'join',
            'description': 'Time-based interval join',
            'performance': '1500M rows/sec'
        })
        
        registry.register('asof_join', CythonAsofJoinOperator, metadata={
            'type': 'join',
            'description': 'Point-in-time join (backward/forward)',
            'performance': '22,221M rows/sec'
        })
        
        logger.info("Registered join operators")
        
    except ImportError as e:
        logger.warning(f"Failed to import join operators: {e}")
    
    # Register window operators
    try:
        from sabot._cython.arrow.window_processor import WindowProcessor
        
        registry.register('window', WindowProcessor, metadata={
            'type': 'window',
            'description': 'Windowed aggregations (tumbling, sliding, session)',
            'performance': 'Variable based on window type'
        })
        
        logger.info("Registered window operators")
        
    except ImportError as e:
        logger.warning(f"Failed to import window operators: {e}")
    
    # Register SQL operators (if available)
    try:
        from sabot.engines.sql.operators import register_sql_operators
        register_sql_operators(registry)
        logger.info("Registered SQL operators")
    except ImportError:
        logger.debug("SQL operators not available")
    
    # Register Graph operators (if available)
    try:
        from sabot.engines.graph.operators import register_graph_operators
        register_graph_operators(registry)
        logger.info("Registered graph operators")
    except ImportError:
        logger.debug("Graph operators not available")
    
    return registry


# Global registry instance (singleton pattern)
_global_registry: Optional[OperatorRegistry] = None


def get_global_registry() -> OperatorRegistry:
    """
    Get the global operator registry.
    
    Creates registry on first access (lazy initialization).
    
    Returns:
        OperatorRegistry: Global registry instance
    """
    global _global_registry
    
    if _global_registry is None:
        _global_registry = create_default_registry()
    
    return _global_registry


def reset_global_registry():
    """Reset global registry (mainly for testing)."""
    global _global_registry
    _global_registry = None

File: sabot_core/src/operators/registry.cpp
#include "sabot/operators/registry.h"
#include <algorithm>
#include <chrono>

namespace sabot {
namespace operators {

OperatorRegistry::OperatorRegistry() 
    : lookup_count_(0), total_lookup_ns_(0) {
}

void OperatorRegistry::Register(
    const std::string& name,
    OperatorType type,
    const OperatorMetadata& metadata
) {
    operator_map_[name] = type;
    
    // Store metadata
    OperatorMetadata meta = metadata;
    meta.name = name;
    meta.type = type;
    metadata_map_[name] = meta;
}

OperatorType OperatorRegistry::Lookup(const std::string& name) const {
    #ifdef SABOT_PROFILE_REGISTRY
    auto start = std::chrono::high_resolution_clock::now();
    #endif
    
    auto it = operator_map_.find(name);
    
    #ifdef SABOT_PROFILE_REGISTRY
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start);
    lookup_count_++;
    total_lookup_ns_ += duration.count();
    #endif
    
    if (it != operator_map_.end()) {
        return it->second;
    }
    
    return OperatorType::CUSTOM;
}

bool OperatorRegistry::HasOperator(const std::string& name) const {
    return operator_map_.find(name) != operator_map_.end();
}

const OperatorMetadata* OperatorRegistry::GetMetadata(const std::string& name) const {
    auto it = metadata_map_.find(name);
    if (it != metadata_map_.end()) {
        return &it->second;
    }
    return nullptr;
}

std::vector<std::string> OperatorRegistry::ListOperators() const {
    std::vector<std::string> names;
    names.reserve(operator_map_.size());
    
    for (const auto& pair : operator_map_) {
        names.push_back(pair.first);
    }
    
    // Sort for consistent ordering
    std::sort(names.begin(), names.end());
    
    return names;
}

OperatorRegistry::Stats OperatorRegistry::GetStats() const {
    Stats stats;
    stats.num_operators = operator_map_.size();
    stats.num_lookups = lookup_count_;
    
    if (lookup_count_ > 0) {
        stats.avg_lookup_ns = static_cast<double>(total_lookup_ns_) / lookup_count_;
    }
    
    return stats;
}

void OperatorRegistry::Clear() {
    operator_map_.clear();
    metadata_map_.clear();
    lookup_count_ = 0;
    total_lookup_ns_ = 0;
}

std::unique_ptr<OperatorRegistry> CreateDefaultRegistry() {
    auto registry = std::make_unique<OperatorRegistry>();
    
    // Register transform operators
    OperatorMetadata filter_meta;
    filter_meta.description = "Filter rows based on predicate";
    filter_meta.performance_hint = "10-500M rows/sec (SIMD)";
    filter_meta.is_stateful = false;
    filter_meta.requires_shuffle = false;
    registry->Register("filter", OperatorType::FILTER, filter_meta);
    
    OperatorMetadata map_meta;
    map_meta.description = "Transform rows with function";
    map_meta.performance_hint = "10-100M rows/sec";
    map_meta.is_stateful = false;
    map_meta.requires_shuffle = false;
    registry->Register("map", OperatorType::MAP, map_meta);
    
    OperatorMetadata select_meta;
    select_meta.description = "Project columns (zero-copy)";
    select_meta.performance_hint = "50-1000M rows/sec";
    select_meta.is_stateful = false;
    select_meta.requires_shuffle = false;
    registry->Register("select", OperatorType::SELECT, select_meta);
    
    OperatorMetadata flat_map_meta;
    flat_map_meta.description = "1-to-N row expansion";
    flat_map_meta.performance_hint = "5-50M rows/sec";
    flat_map_meta.is_stateful = false;
    flat_map_meta.requires_shuffle = false;
    registry->Register("flat_map", OperatorType::FLAT_MAP, flat_map_meta);
    
    OperatorMetadata union_meta;
    union_meta.description = "Merge multiple streams";
    union_meta.performance_hint = "5-50M rows/sec";
    union_meta.is_stateful = false;
    union_meta.requires_shuffle = false;
    registry->Register("union", OperatorType::UNION, union_meta);
    
    // Register aggregation operators
    OperatorMetadata group_by_meta;
    group_by_meta.description = "Group by keys and aggregate";
    group_by_meta.performance_hint = "5-100M rows/sec";
    group_by_meta.is_stateful = true;
    group_by_meta.requires_shuffle = true;
    registry->Register("group_by", OperatorType::GROUP_BY, group_by_meta);
    
    OperatorMetadata aggregate_meta;
    aggregate_meta.description = "Global aggregation";
    aggregate_meta.performance_hint = "2-50M rows/sec";
    aggregate_meta.is_stateful = true;
    aggregate_meta.requires_shuffle = false;
    registry->Register("aggregate", OperatorType::AGGREGATE, aggregate_meta);
    
    OperatorMetadata distinct_meta;
    distinct_meta.description = "Deduplication";
    distinct_meta.performance_hint = "1-10M rows/sec";
    distinct_meta.is_stateful = true;
    distinct_meta.requires_shuffle = true;
    registry->Register("distinct", OperatorType::DISTINCT, distinct_meta);
    
    // Register join operators
    OperatorMetadata hash_join_meta;
    hash_join_meta.description = "Hash join on keys";
    hash_join_meta.performance_hint = "104M rows/sec";
    hash_join_meta.is_stateful = true;
    hash_join_meta.requires_shuffle = true;
    registry->Register("hash_join", OperatorType::HASH_JOIN, hash_join_meta);
    
    OperatorMetadata interval_join_meta;
    interval_join_meta.description = "Time-based interval join";
    interval_join_meta.performance_hint = "1500M rows/sec";
    interval_join_meta.is_stateful = true;
    interval_join_meta.requires_shuffle = false;
    registry->Register("interval_join", OperatorType::INTERVAL_JOIN, interval_join_meta);
    
    OperatorMetadata asof_join_meta;
    asof_join_meta.description = "Point-in-time join";
    asof_join_meta.performance_hint = "22,221M rows/sec";
    asof_join_meta.is_stateful = true;
    asof_join_meta.requires_shuffle = false;
    registry->Register("asof_join", OperatorType::ASOF_JOIN, asof_join_meta);
    
    // Register window operator
    OperatorMetadata window_meta;
    window_meta.description = "Windowed aggregations";
    window_meta.performance_hint = "Variable based on window type";
    window_meta.is_stateful = true;
    window_meta.requires_shuffle = false;
    registry->Register("window", OperatorType::WINDOW, window_meta);
    
    return registry;
}

// Global singleton
static std::unique_ptr<OperatorRegistry> global_registry = nullptr;

OperatorRegistry& GetGlobalRegistry() {
    if (!global_registry) {
        global_registry = CreateDefaultRegistry();
    }
    return *global_registry;
}

} // namespace operators
} // namespace sabot

